// @sentry/electron/renderer/index.js
const log = require('./log.js');
const browser = require('@sentry/browser');
const scopeToMain = require('./integrations/scope-to-main.js');
const eventLoopBlock = require('./integrations/event-loop-block.js');
const transport = require('./transport.js');
const stackParse = require('./stack-parse.js');
const sdk = require('./sdk.js');



exports.logger = log;
exports.BrowserClient = browser.BrowserClient;
exports.OpenFeatureIntegrationHook = browser.OpenFeatureIntegrationHook;
exports.SEMANTIC_ATTRIBUTE_SENTRY_OP = browser.SEMANTIC_ATTRIBUTE_SENTRY_OP;
exports.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN = browser.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE = browser.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE = browser.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE;
exports.Scope = browser.Scope;
exports.addBreadcrumb = browser.addBreadcrumb;
exports.addEventProcessor = browser.addEventProcessor;
exports.addIntegration = browser.addIntegration;
exports.breadcrumbsIntegration = browser.breadcrumbsIntegration;
exports.browserApiErrorsIntegration = browser.browserApiErrorsIntegration;
exports.browserProfilingIntegration = browser.browserProfilingIntegration;
exports.browserSessionIntegration = browser.browserSessionIntegration;
exports.browserTracingIntegration = browser.browserTracingIntegration;
exports.buildLaunchDarklyFlagUsedHandler = browser.buildLaunchDarklyFlagUsedHandler;
exports.captureConsoleIntegration = browser.captureConsoleIntegration;
exports.captureEvent = browser.captureEvent;
exports.captureException = browser.captureException;
exports.captureFeedback = browser.captureFeedback;
exports.captureMessage = browser.captureMessage;
exports.captureSession = browser.captureSession;
exports.chromeStackLineParser = browser.chromeStackLineParser;
exports.consoleLoggingIntegration = browser.consoleLoggingIntegration;
exports.contextLinesIntegration = browser.contextLinesIntegration;
exports.continueTrace = browser.continueTrace;
exports.createTransport = browser.createTransport;
exports.createUserFeedbackEnvelope = browser.createUserFeedbackEnvelope;
exports.dedupeIntegration = browser.dedupeIntegration;
exports.defaultRequestInstrumentationOptions = browser.defaultRequestInstrumentationOptions;
exports.endSession = browser.endSession;
exports.eventFiltersIntegration = browser.eventFiltersIntegration;
exports.eventFromException = browser.eventFromException;
exports.eventFromMessage = browser.eventFromMessage;
exports.exceptionFromError = browser.exceptionFromError;
exports.extraErrorDataIntegration = browser.extraErrorDataIntegration;
exports.featureFlagsIntegration = browser.featureFlagsIntegration;
exports.feedbackAsyncIntegration = browser.feedbackAsyncIntegration;
exports.feedbackIntegration = browser.feedbackIntegration;
exports.feedbackSyncIntegration = browser.feedbackSyncIntegration;
exports.forceLoad = browser.forceLoad;
exports.functionToStringIntegration = browser.functionToStringIntegration;
exports.getActiveSpan = browser.getActiveSpan;
exports.getClient = browser.getClient;
exports.getCurrentScope = browser.getCurrentScope;
exports.getFeedback = browser.getFeedback;
exports.getGlobalScope = browser.getGlobalScope;
exports.getIsolationScope = browser.getIsolationScope;
exports.getReplay = browser.getReplay;
exports.getRootSpan = browser.getRootSpan;
exports.getSpanDescendants = browser.getSpanDescendants;
exports.getSpanStatusFromHttpCode = browser.getSpanStatusFromHttpCode;
exports.getTraceData = browser.getTraceData;
exports.globalHandlersIntegration = browser.globalHandlersIntegration;
exports.graphqlClientIntegration = browser.graphqlClientIntegration;
exports.httpClientIntegration = browser.httpClientIntegration;
exports.httpContextIntegration = browser.httpContextIntegration;
exports.inboundFiltersIntegration = browser.inboundFiltersIntegration;
exports.instrumentOutgoingRequests = browser.instrumentOutgoingRequests;
exports.instrumentSupabaseClient = browser.instrumentSupabaseClient;
exports.isEnabled = browser.isEnabled;
exports.isInitialized = browser.isInitialized;
exports.lastEventId = browser.lastEventId;
exports.launchDarklyIntegration = browser.launchDarklyIntegration;
exports.linkedErrorsIntegration = browser.linkedErrorsIntegration;
exports.moduleMetadataIntegration = browser.moduleMetadataIntegration;
exports.onLoad = browser.onLoad;
exports.openFeatureIntegration = browser.openFeatureIntegration;
exports.parameterize = browser.parameterize;
exports.registerSpanErrorInstrumentation = browser.registerSpanErrorInstrumentation;
exports.registerWebWorker = browser.registerWebWorker;
exports.replayCanvasIntegration = browser.replayCanvasIntegration;
exports.replayIntegration = browser.replayIntegration;
exports.reportingObserverIntegration = browser.reportingObserverIntegration;
exports.rewriteFramesIntegration = browser.rewriteFramesIntegration;
exports.sendFeedback = browser.sendFeedback;
exports.setContext = browser.setContext;
exports.setCurrentClient = browser.setCurrentClient;
exports.setExtra = browser.setExtra;
exports.setExtras = browser.setExtras;
exports.setHttpStatus = browser.setHttpStatus;
exports.setMeasurement = browser.setMeasurement;
exports.setTag = browser.setTag;
exports.setTags = browser.setTags;
exports.setUser = browser.setUser;
exports.showReportDialog = browser.showReportDialog;
exports.spanToBaggageHeader = browser.spanToBaggageHeader;
exports.spanToJSON = browser.spanToJSON;
exports.spanToTraceHeader = browser.spanToTraceHeader;
exports.spotlightBrowserIntegration = browser.spotlightBrowserIntegration;
exports.startBrowserTracingNavigationSpan = browser.startBrowserTracingNavigationSpan;
exports.startBrowserTracingPageLoadSpan = browser.startBrowserTracingPageLoadSpan;
exports.startInactiveSpan = browser.startInactiveSpan;
exports.startNewTrace = browser.startNewTrace;
exports.startSession = browser.startSession;
exports.startSpan = browser.startSpan;
exports.startSpanManual = browser.startSpanManual;
exports.statsigIntegration = browser.statsigIntegration;
exports.supabaseIntegration = browser.supabaseIntegration;
exports.suppressTracing = browser.suppressTracing;
exports.thirdPartyErrorFilterIntegration = browser.thirdPartyErrorFilterIntegration;
exports.unleashIntegration = browser.unleashIntegration;
exports.updateSpanName = browser.updateSpanName;
exports.webWorkerIntegration = browser.webWorkerIntegration;
exports.withActiveSpan = browser.withActiveSpan;
exports.withIsolationScope = browser.withIsolationScope;
exports.withScope = browser.withScope;
exports.zodErrorsIntegration = browser.zodErrorsIntegration;
exports.scopeToMainIntegration = scopeToMain.scopeToMainIntegration;
exports.eventLoopBlockIntegration = eventLoopBlock.eventLoopBlockIntegration;
exports.makeRendererTransport = transport.makeRendererTransport;
exports.defaultStackParser = stackParse.electronRendererStackParser;
exports.getDefaultIntegrations = sdk.getDefaultIntegrations;
exports.init = sdk.init;
